driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -280,21 +268,37 @@ public void doMaintenance() {
         }
     }
 
-    private PooledConnection getPooledConnection(final long timeout, final TimeUnit timeUnit) {
-        UsageTrackingInternalConnection internalConnection = pool.get(timeout, timeUnit);
-        while (shouldPrune(internalConnection)) {
-            pool.release(internalConnection, true);
-            internalConnection = pool.get(timeout, timeUnit);
-        }
-        if (LOGGER.isTraceEnabled()) {
-            LOGGER.trace(format("Checked out connection [%s] to server %s", getId(internalConnection), serverId.getAddress()));
+    private PooledConnection getPooledConnection(final Timeout timeout) throws MongoTimeoutException {
+        boolean externalTimeoutException = false;
+        try {
+            UsageTrackingInternalConnection internalConnection = pool.get(timeout.remainingNanosOrInfinite(), TimeUnit.NANOSECONDS);
+            while (shouldPrune(internalConnection)) {
+                pool.release(internalConnection, true);
+                long remainingNanos = timeout.remainingNanosOrInfinite();
+                if (Timeout.expired(remainingNanos)) {
+                    externalTimeoutException = true;
+                    throw createTimeoutException(timeout);
+                }
+                internalConnection = pool.get(remainingNanos, TimeUnit.NANOSECONDS);
+            }
+            if (LOGGER.isTraceEnabled()) {
+                LOGGER.trace(format("Got connection [%s] to server %s", getId(internalConnection), serverId.getAddress()));
+            }
+            return new PooledConnection(internalConnection);
+        } catch (MongoTimeoutException e) {
+            if (externalTimeoutException) {
+                throw e;
+            } else {
+                MongoTimeoutException timeoutException = createTimeoutException(timeout);
+                timeoutException.initCause(e);
```

Comment:
Note that I used `addSuppressed` here previously, but now I use `initCause`.

Reply:
Here I follow a code pattern that results in not losing potentially useful information even when facing future code changes: unless the caught exception must not be externally observable, propagate it. This way I do not need to think each time to what extent the information in the caught exception is useful, and whatever the result of my speculation is, the exception handling stays correct regardless of the future changes of the code that throws the caught exception.

It is also worth pointing out the the method `Throwable.initCause` is used here with accordance to its design: it is used to specify the cause when creating an exception that does not have other support for setting the cause.

Do you see a problem with this code?

Reply:
It might be confusing to users when printed out in a log file to see MongoTimeoutException, caused by MongoTimeoutException.

Reply:
Convinced) Done.